### PR TITLE
lvm.nback support for named alloca

### LINKED ISSUE #56: llvm.nback should make use of name argument to alloca
These correspond to local variables, so there's an obvious semantic.

Follow on from #55.

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
lvm.nback should make use of name argument to alloca
Resolves #56 

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: compiler/modules/nback.llvm/llvm.bal [hunk 1]
```diff
@@ -179,7 +179,7 @@ public class FunctionDefn {
     final string functionName;
 
     private BasicBlock[] basicBlocks = [];
-    private int varCount = 0;
+    private string[] variableNames = [];
```
Comment (jclark) [inline compiler/modules/nback.llvm/llvm.bal]: Don't think this the right data structure.

Comment (jclark) [reply compiler/modules/nback.llvm/llvm.bal]: You should have a map from variable name to count.

File: compiler/modules/nback.llvm/llvm.bal [hunk 2]
```diff
@@ -235,9 +235,18 @@ public class FunctionDefn {
         return label;
     }
 
-    function genReg() returns string {
-        string reg = "%" + "R" + self.varCount.toString();
-        self.varCount += 1;
+    function genReg(string? name = ()) returns string {
```
Comment (jclark) [inline compiler/modules/nback.llvm/llvm.bal:238]: To be completely bullet proof here, you need deal with the case where the name passed to genReg ends  with a `.`  followed by a number

File: compiler/modules/nback.llvm/llvm.bal [hunk 3]
```diff
@@ -235,9 +235,18 @@ public class FunctionDefn {
...
+        string regName = "R" + self.variableNames.length().toString();
```
Comment (jclark) [inline compiler/modules/nback.llvm/llvm.bal]: If we are now using user-defined names, then `R` is no longer a choice (it looks too much like a user variable). Something like `_` would be better if it works.

File: compiler/modules/nback.llvm/llvm.bal [hunk 4]
```diff
@@ -531,6 +540,19 @@ function createLine(string[] words, string indent = "") returns string {
     return string:concat(indent, ...parts);
 }
 
+function incrementName(string name, int count) returns string{
+    if count == 0 {
+       return name + "." + count.toString(); 
+    } else {
+        int? splitPoint = name.lastIndexOf(".");
```
Comment (jclark) [inline compiler/modules/nback.llvm/llvm.bal]: In this sort of case, just do `int splitPoint = <int>name.lastIndexOf(".");`

But I don't think you need this with the right data structure.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/nback.llvm/llvm.bal (hunk 4 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback.llvm/llvm.bal]: Fixed by changing variable name store to map

File: compiler/modules/nback.llvm/llvm.bal (hunk 3 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback.llvm/llvm.bal]: Fixed

File: compiler/modules/nback.llvm/llvm.bal (hunk 1 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback.llvm/llvm.bal]: Fixed

File: compiler/modules/nback.llvm/llvm.bal (hunk 2 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback.llvm/llvm.bal:238]: Fixed

**Pull request merged**